src/index.d.ts

```diff
@@ -0,0 +1,58 @@
+declare type elementGenerator = (
+	stream: Renderer.Stream,
+	nonBlocking: boolean,
+	callback: () => void
+) => void;
+
+interface StatelessFunctionalComponent<P> {
+	(props: P): elementGenerator;
+}
+
+declare function createElement<T>(
+	element: string | StatelessFunctionalComponent<T>,
+	params: T,
+	...children
+): elementGenerator;
+
+declare function safe(str: string): Renderer.HTMLString;
+
+declare class Renderer {
+	constructor(doctype?: string);
+
+	renderView(
+		view: elementGenerator | string,
+		params: Object,
+		stream: Renderer.Stream,
+		{ fragment: boolean }?,
+		callback?: () => void
+	): void;
+
+	registerView(
+		macro: () => elementGenerator,
+		name?: string,
+		replace?: boolean
+	): string;
+}
+
+declare namespace Renderer {
+	export interface Stream {
+		write(msg: string): void;
+		writeln(msg: string): void;
+		flush(): void;
+	}
+	export class HTMLString {
+
+	}
+}
+
+export default Renderer;
+
+export { createElement, safe };
```

Comment:
I prefer for a module's public API to reside at the top - can we safely move this accordingly, would forward-references be hoisted?

Reply:
done